### DISCUSSION TITLE
Proof of Concept

### DISCUSSION THREAD

**Discussion by Jeewhan** [InnerCircleA/design-system#6] [General]
미션 내용
1. 팀 빌드업 및 역할 구성
2. 프로토타입 목표 설정
3. 프로토타입 구현

**Comment by Jeewhan** [comment]
컴포넌트 기반 디자인 시스템 구축 목표(https://retool.com/) 로 합의.

아래는 카카오톡으로 제안한 사항 요약

1. 각자 12/18(일)까지 PoC를 수행해서 이해도를 높인 뒤 그동안의 Lesson Learn에 대해 공유하고 논의하기.
  1-1. 이 과정에서 코드의 퀄리티 등을 따지기 보다는 플로우에 대한 검증을 목표.

2. 컴포넌트 기반의 디자인 시스템을 구축하기 위해 필요한 전체 플로우에 대한 가안을 미팅을 통해 함께 그리고, 각자 관심있는 요소들을 PoC 대상으로 선택. (아래는 플로우 내 요소 예시)
- Figma 플러그인을 통한 HTML & CSS 추출
- Editor를 통한 Preview 또는 Edit
- Figma 플러그인 결과물을 반영한 디자인 컴포넌트 설계.
- Preview 등을 위한 API & DB
- 프로덕트 내 디자인 컴포넌트 사용처 추적을 위한 Webpack 플러그인 또는 TypeScript AST Parser

3. 요소별 담당자를 두되 요소 간 인터페이스를 잘 정의하고 추후에 모노레포를 엮는 것을 통해 최대한 서로에 대한 작업의존성이 낮은 형태로 협업을 수행하되 코드 리뷰는 수행.

4. PoC를 수행하기 전에 PoC 킥오프 미팅을 통해 전체 플로우에 대한 가안을 함께 그리기를 제안.

=> **전체 플로우에 대한 가안을 그리는 과정을 PoC 이후에 진행하자는 @gyutae100 님의 의견을 반영하여 그대로 진행.**

---

임의로 작성한 플로우 가안

![image](https://user-images.githubusercontent.com/14256139/207026277-e448ec6e-58e8-46e6-b22e-e36e420b105d.png)

**Reply by peanudge** [reply]
<img width="366" alt="스크린샷 2022-12-12 오후 9 39 13" src="https://user-images.githubusercontent.com/20085849/207047349-30139561-43ef-441f-bc2d-fbe4afb29454.png">

아마 웹 컴포넌트에서 변환은 빠져도 괜찮을 것 같습니다. 규태님과 소통했을 때 **웹컴포넌트**는 저희 관심사에서 사라졌었던 것 같아요. `JSON으로 리액트 컴포넌트를 만든다.`로 축약해도 괜찮을 것 같아요.

**Reply by peanudge** [reply]
> Preview 등을 위한 API & DB

그리고 Preview 자체가 꽤 귀찮은 작업이 될 수도 있는데 만들어진 컴포넌트는 스토리북을 활용하면 간편하고 유용할 것 같아요.

**Reply by Jeewhan** [reply]
> Preview 자체가 꽤 귀찮은 작업이 될 수도 있는데 만들어진 컴포넌트는 스토리북을 활용하면 간편하고 유용할 것 같아요.

@peanut-lover 네네, 좋습니다. 다만 컴포넌트 사용처 등의 정보를 스토리북에서 보여줄 수 있을지에 대한 확인은 필요할 듯 싶습니다. 그 부분까지 담을 수 있다면 저희의 관심사와 리소스를 감안했을 때 탁월한 선택지가 될 것 같습니다!

**Reply by Jeewhan** [reply]
> 아마 웹 컴포넌트에서 변환은 빠져도 괜찮을 것 같습니다. 규태님과 소통했을 때 웹컴포넌트는 저희 관심사에서 사라졌었던 것 같아요. JSON으로 리액트 컴포넌트를 만든다.로 축약해도 괜찮을 것 같아요.

@peanut-lover 네네, 좋습니다. 한 가지 고민이 되었던 점은 (위의 스토리북과도 이어지는데) 특정 외부 라이브러리(리액트, 스토리북 등등)에 대한 의존성을 적절하게 격리시킬 수 있느냐 였습니다.

리액트 특성상 그게 쉽진 않을 것 같기도 합니다... (물론 애를 쓰면 비즈니스 로직은 최대한 순수하게 작성하고 리액트가 하는 일은 최대한 바인딩만 하는 형태로 구현할 수는 있겠지만 그래도 상태가 엮이면 쉽진 않을 듯 합니다.)

그럼에도 저희에게 주어진 기간과 리소스가 한정되어있기 때문에 지금은 리액트나 스토리북과 같이 의존할 수 있는 외부 요소를 최대한 적극적으로 활용하는게 좋을 것 같습니다.

**Reply by peanudge** [reply]
간단한 프로토타입에서는 리액트 의존해도 괜찮을 것 같아요.

**Reply by gyutae100** [reply]
지환님이 가독성 높게 도식화 해주면서 설명해주셔서, 덕분에 쉽게 팔로우 업 했어요.
같이 고민하고 작성할 부분인데, 주도적으로 작성해주셔서 특히 감사합니다.

그리고 두분께서 좋은 논의를 해주신 덕분에, 기존에 가진 모호한 부분이 상당 부분 해소 됬어요.

두분께서 진도를 많이 진행 된걸로 판단되는데, 저도 어서 부지런히 쫒아가야겠어요.

**Comment by peanudge** [comment]
# Component Tracking Plugin PoC 요약

Tracking Plugin 설계에 대한 1차적인 목표인 `어떤 컴포넌트를 어디서 사용했는지 확인할 수 있다`를 달성해서 PoC 구현체와 과정을 `poc/wepback-tracking-plugin` 에 README 파일로 남겨놨습니다.

https://github.com/InnerCircleA/design-system/tree/main/webpack-tracking-plugin/poc

## Poc 결과 정리

- 저희 디자인 시스템에 도달하기 위해서 기술적으로 **프로젝트별 컴포넌트 사용 트레킹**이 가능성 확인
- `project` 위치, `page` 위치, 사용된 `component` 종류 데이터를 회수 중이지만 그외에도 **의미있는 데이터(프로젝트 정보)가 많음**을 확인

## 추가적으로 해볼만한 작업

- 사용된 컴포넌트의 버젼 트레킹을 할 수 있어야한다.
- 위 정보들을 한 곳에 모아 모니터링할 수 있을 데이터 구조 설립
- 컴포넌트 트레킹을 실제 프로젝트에 활용해서 부족한 부분 확인

**Comment by Jeewhan** [comment]
월요일(2022-12-19) 미팅후 정해진 플로우 및 아키텍처 초안 (표기된 섹션은 진행한 PoC 범위 및 수행자)

향후 아래 요소에 대응되는 각각의 Repository들을 Monorepo를 활용해 구현하는 형태 예상.

![image](https://user-images.githubusercontent.com/14256139/208445373-155fa0d0-d291-423f-9acb-dbe9fd4c6573.png)

**Comment by Jeewhan** [comment]
# Figma Export PoC 요약

`Figma로부터 데이터 가져오기` 와 관련하여 PoC를 진행하였습니다.
https://github.com/InnerCircleA/design-tokens

## PoC 결과 정리

- 디자인 시스템에 필요한 design foundation을 Figma로부터 design token을 통해 얻어올 수 있었습니다.
- design token를 활용해 디자인 시스템 컴포넌트의 스타일 작성을 위한 css-in-js에 사용될 JS variables를 만들어낼 수 있었습니다.
- Figma -> Tokens Studio -> token-transformer -> Style Dictionary -> JS variables
- 아직 진행하지 못했지만, 이후에 px를 붙여주거나 숫자 반올림 처리 등이 필요할 것으로 보입니다.

## 이어서 진행해볼 수 있는 선택지들

- design token 관련 TypeScript
- Github push되었을 때 후처리에 대한 GIthub Actions와 버전 관리
- Figma Plugin API를 활용하여 component export
- Figma Widget API를 활용하여 디자이너를 위한 style lint
- ...


## PoC를 진행한 뒤 생겨난 고민들

- 무엇이 우리 디자인 시스템의 킬링피처가 되어야 하는가?
- 필요하다면 디자인시스템 관련 사용자(플랫폼 디자이너, UX 엔지니어 등) 인터뷰를 진행해볼 수도 있을텐데 인터뷰에 대한 질문 설계를 어떻게 해야 할 것인가?

**Reply by peanudge** [reply]
> 필요하다면 디자인시스템 관련 사용자(플랫폼 디자이너, UX 엔지니어 등) 인터뷰를 진행해볼 수도 있을텐데 인터뷰에 대한 질문 설계를 어떻게 해야 할 것인가?

위 고민에 대해서는 저희에게 **중요한 고민**인 것에 공감합니다. 사실 PoC 주제 선정까지만 봐도 저희는 엔지니어 관점에서 디자인 시스템을 위한 툴에 중점을 두는 것 같아요. 그래서 디자인 시스템과 관련된 업무를 하는 타사 개발자인 지인들의 피드백을 받았었는데

>우리 디자인 시스템의 킬링피처

`디자인 시스템이 제대로 동작할 수 있도록 지원해줄 수 있는 툴을 만들고 싶다`가 제가 생각하는 킬링 피쳐의 방향성이지만 이에대한 지인들은 `디자인 시스템이 아닌 디자인 시스템을 도와주는 툴을 개발하는거 아니냐`는 피드백을 받았어요. 우리의 방향이 디자이너 관점의 디자인 시스템이 많이 빠진 느낌이라고 생각이 들기도 했습니다. 컴포넌트 디자인에 대한 제약, 확장에 대한 철학 같은 것을 고민해볼 시점은 언제쯤 될까싶어요.
